src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -4959,6 +4959,37 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
             handleMissing
           }
           else {
+            if ((sym ne NoSymbol) && !qual.tpe.isStable && argsDependOnPrefix(sym)) {
+              // Rewrites "qual.name ..." to "{ val lhs = qual ; lhs.name ... }" in cases where
+              // qual is not stable and name has a method type which depends on its prefix. If
+              // this is the case then hoisting qual out as a stable val means that members of
+              // implicit scopes which are accessible via lhs can be candidates for satisfying
+              // implicit (conversions to) arguments of name.
+
+              // We have to introduce the ValDef before its use here, so we walk up the
+              // context tree and attach it to the original root of this expression. It will
+              // be extracted and inserted by insertStabilizer when typer unwinds out of this
+              // expression.
+              val insertionContext = context.nextEnclosing { ctx =>
+                def isInsertionNode(tree: Tree): Boolean = tree match {
+                  case _: Apply | _: TypeApply | _: Select => true
+                  case _ => false
+                }
+                isInsertionNode(ctx.tree) && !isInsertionNode(ctx.outer.tree)
+              }
+
+              if(insertionContext != NoContext) {
+                val vsym = insertionContext.owner.newValue(unit.freshTermName(nme.STABILIZER_PREFIX), qual.pos.focus, SYNTHETIC | ARTIFACT | STABLE)
+                vsym.setInfo(qual.tpe)
```

Comment:
I think we might have the same issue as https://github.com/scala/bug/issues/7694 here, fixed in named / default args with the `uncheckedBounds` annotation (https://github.com/scala/scala/blob/v2.13.0-M2/src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala#L173-L189)

```
scala> :pa -raw
// Entering paste mode (ctrl-D to finish)

trait A
trait B

class Foo[A2, B2 <: A2] {
  class Bar
  object Bar {
    implicit def fromString(a: String) = new Bar
  }
  def andThen(b : Bar) = b
}

object Test {
  def lub = if (true) (null: Foo[A, A]) else (null: Foo[B, B])
  (lub) andThen ("Bar")
}

// Exiting paste mode, now interpreting.


[[syntax trees at end of                     typer]] // <pastie>

    def lub: Foo[_ >: A with B <: Object, _ >: A with B <: Object] = if (true)
      (null: Foo[A,A])
    else
      (null: Foo[B,B]);
    {
      <synthetic> <stable> <artifact> val stabilizer$1: Foo[_1,_2] = Test.this.lub;
      stabilizer$1.andThen(stabilizer$1.Bar.fromString("Bar"))
    }


         (lub) andThen ("Bar")
          ^
<pastie>:14: error: type arguments [_1,_2] do not conform to class Foo's type parameter bounds [A2,B2 <: A2]
There were compilation errors!
```

Reply:
Thanks for the pointer.